### PR TITLE
Move functorch decompositions to PyTorch

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76311

This moves all of the preexisting functorch decompositions from https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py into PyTorch. I ended up rabbitholing on testing strategy, so there are actually some substantive changes, which I will describe shortly.

If you are **not** familiar with the preexisting functorch decompositions, here is what's in this PR.

* API for maintaining a global table of decompositions (mapping of OpOverload to callable) in `torch/_decomp/__init__.py`
* A decently large number of decompositions in `torch/_decomp/decompositions.py`
* A bug fix for `hasattr(torch.ops.aten, "blah")` which incorrectly raises a RuntimeError instead of returning False (we need to appropriately raise an AttributeError); this was previously worked around in functorch by explicitly nulling out the `__origin__` attribute on namespaces--this is no longer necessary
* A new `aten_backward_name` field on OpInfos, recording the name of the aten backwards operator, analogous to `aten_name` (which records the name of the aten forwards operator)
* Tests for the decompositions in `test_decomp.py`. The test structure is to run the forwards and then backwards (here, I faithfully reused `ref_vjp_no_create` from functorch), and do this while running a mode that, whenever it hits a decomposable op, runs both the original and the decomposition and checks them if they're close enough. On half precision operations, we check that the result is as good or closer to double precision result; on other precisions we just test for normal equality. There are some skips for known broken cases involving complex numbers as well as some precision problems. `test_quick` only runs tests whose `aten_name` or `aten_backward_name` are in the decomposition table; `test_comprehensive` runs everything but is a lot slower to run.

Some decompositions are not exercised on CPU with the stock PyTorch OpInfo database. In particular, the omission of `native_dropout` and `clamp_max` tests seems problematic.

```
aten._fused_dropout.default
aten.cudnn_batch_norm.default
aten.native_dropout.default
aten.native_dropout_backward.default
aten.clamp_max.default
aten.cudnn_batch_norm_backward.default
```

The detach decomposition is a bit questionable; an alias call should be sufficient (and these should disappear entirely if functionalization was applied).

If you are familiar with the functorch decompositions:

* There were many fixes in both helper functions in decompositions to handle complex numbers
* `_autograd_grad` was updated, c.f., https://github.com/pytorch/functorch/pull/750
* `op_assert_ref` was updated to work for boolean tensors
* `op_assert_equal` is updated to just use `assertEqual` for the is close test. This makes it work for half tensors on CPU.
* The test function no longer uses decomposition tensor wrapper; instead a mode is used to interpose on decompositions.
* Previously the test harness was responsible for converting float16/bfloat16 inputs to float32 before running the decompositions; now this logic lives in the decomps itself as `cast_for_opmath` (this decorator isn't 100% correct in the presence of type promotion but it is sufficient for all the OpInfo tests--let's improve it some more later)
* The handling for `do_relative_check` is refactored to have two separate for loops
* The handling for dtype eligibility is improved as we now appropriately test `supported_backward_dtypes` before attempting to run backwards
* We test to make sure a decomposition is actually run in the quick tests (thus, the quick tests are not 100% subsumed by the comprehensive tests, which don't bother to see if a decomp actually ran or not)
* The threshold decomposition did not handle scalar tensors correctly, this has been fixed

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -231,6 +238,36 @@ def test_reference_testing(self, device, dtype, op):
...
+                # TODO: this no_dispatch call shouldn't be necessary after
+                # https://github.com/pytorch/pytorch/pull/75965
+                with no_dispatch():
+                    if func in decomposition_table:
+                        nonlocal hit
+                        hit = True
+                        return decomposition_table[func](*args, **kwargs)
+                    else:
+                        return func(*args, **kwargs)
+
```
Comment (zou3519) [inline test/test_ops.py:261]: @Chillee wrote tests for decompositions over at https://github.com/pytorch/functorch/blob/955decd01aa7fc391f2bfbbb7ae5de0f9ed8ed0a/test/test_ops.py#L1349, should we just upstream those? These cover autograd testing and some additional dtypes.

File: torch/_decomp/__init__.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    def decomposition_decorator(f):
+        nonlocal registry
+        if registry is None:
+            registry = decomposition_table
+
+        def add_op_to_table(aten_op):
+            # Converts aten.foo to aten.foo.default
+            # Done so I can be lazy and not write default on all of these ops
+            if not isinstance(aten_op, torch._ops.OpOverload):
```
Comment (zou3519) [inline torch/_decomp/__init__.py:36]: nit: Could this be confusing? I'm worried someone can interpret aten.foo as "this is the decomposition for all overloads of aten.foo"

**Review by zou3519 — COMMENTED** [review]
Lint failures look legit. I want to know what the testing plan is: there is some decomposition testing in functorch that handles autograd and other devices that could be upstreamed here (is that coming later?). Is this a candidate for cross ref testing?

File: torch/_decomp/__init__.py (hunk 2 above)
Comment (ezyang) [reply torch/_decomp/__init__.py:36]: I'm OK with changing this, I was mostly trying to keep in line with the old code as much as possible. @Chillee any thoughts?

Comment (Chillee) [reply torch/_decomp/__init__.py:36]: hmmm.... I think it's potentially fine.

I wouldn't be massively opposed to forcing an explicit use of `default` on every decomposition - just makes things a bit uglier.

Comment (ezyang) [reply torch/_decomp/__init__.py:36]: I feel there is a more fundamental API design question we should resolve one way or another, which is whether or not it's OK to say torch.ops.aten.foo when we mean torch.ops.aten.foo.default. I know @anjali411 hates it. This area generally seems like an extreme attractive nuisance in the API right now. If we're going to keep the conversion, we should commit to "whenever I get an OpOverloadPacket where I expect an OpOverload, I will grab the default overload" and apply it consistently in all our torch.ops handling APIs.

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -629,6 +629,7 @@ def __init__(self,
                                                    # defaults to the value of `supports_forward_ad`
                  check_inplace_batched_forward_grad=None,   # whether to check batched forward grad when doing gradcheck
                                                             # defaults to the value of `check_batched_forward_grad`
+                 check_decomp=False,  # whether to check against torch._decomp
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_methods_invocations.py:632]: My main goal is I want to be able to run tests for all decompositions quickly; I don't want to waste time churning through forward-backwards for everything in the OpInfo, when only twenty of the operators actually do anything nontrivial.

The test cannot easily query if a decomposition exists for an operator because OpInfos deal with torch API, and decompositions are registered against ATen API. I could dump a table mapping torch functions to aten functions (this would be relatively easy to do), but you now have the same problem of keeping the table up to date (though, I suppose, one could have a separate test testing that the table is up to date). WDYT?

[the remark this answers is not in the document]
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:632]: Yeah that all sounds reasonable. A table is interesting but I wouldn't want to make it a requirement today but more like an option later. But don't all the OpInfos specify their ATen names anyway? Are those not usable?

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_methods_invocations.py:632]: Ah, in this case it's because I need to know what the ATen names of the backwards are, but I suppose we can add that instead!

File: torch/_decomp/__init__.py (hunk 2 above)
Comment (anjali411) [reply torch/_decomp/__init__.py:36]: > whether or not it's OK to say torch.ops.aten.foo when we mean torch.ops.aten.foo.default

I don't think it's a good idea. Not every operator is guaranteed to have a default overload. And in many cases like add, default overload is probably not what you want since it corresponds to the following op: `aten::add(Scalar a, Scalar b) -> (Scalar)`

On a more high level, it seems bad to infer the packet as the default overload by default because calling those two operators on the same set of arguments can lead to different result (because of jit overload resolution). 

Does it make sense to register decompositions for the packets instead of the overloads? To put it another way, do we plan to register different decompositions for different overloads?

Comment (ezyang) [reply torch/_decomp/__init__.py:36]: Hypothetically, you could register a decomp for a packet, in which case you are saying "I will do overload resolution myself", but for the most part I expect people to register decomps for specific overloads. It's quite a bit nicer.

**Review by Chillee — APPROVED** [review]
LGTM!

File: test/test_decomp.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+        requires_grad=requires_grad,
+    )
+
+
+CROSS_REF_EXCLUDE_SET = {
+    (
+        "cpu",
+        torch.bfloat16,
+        "nn.functional.layer_norm",
+    ),  # "batch_norm" not implemented for 'BFloat16'
```
Comment (Chillee) [inline test/test_decomp.py:267]: Why is this needed - is it not already covered by OpInfo dtype support?

**Comment by ezyang** [comment]
@zou3519 latest update now incorporates functorch's tests plus improvements. True crossref testing I think will late for later.

File: test/test_decomp.py (hunk 4 above)
Comment (ezyang) [reply test/test_decomp.py:267]: I am not sure but empirically the test failed even after I made us respect backward dtypes.

File: test/test_decomp.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyNativeDeviceTypes
+    @skipIfCrossRef
+    @suppress_warnings
+    @ops(op_db)
+    def test_comprehensive(self, device, dtype, op):
+        self.do_cross_ref(device, dtype, op, run_all=True)
+
+    def do_cross_ref(self, device, dtype, op, *, run_all):
+        if (torch.device(device).type, dtype, op.name) in CROSS_REF_EXCLUDE_SET or (
```
Comment (mruberry) [inline test/test_decomp.py:330]: fyi maybe these skips could be modeled as skips on the OpInfos themselves for consistency?

Comment (mruberry) [reply test/test_decomp.py:330]: I guess the issue with that is you'd have to skip multiple tests, which is kinda a pain?

Comment (ezyang) [reply test/test_decomp.py:330]: I thought about it... it just seemed a lot harder to handle the skips if they were smeared out through the OpInfo database, rather than centralized on the test itself. I guess this is a general point on OpInfo skips: I think they should be attached to the relevant tests, rather than the OpInfo entries themselves.

**Comment by ezyang** [comment]
closed by https://github.com/pytorch/pytorch/commit/a3f10ec28176b8bda21e54ace5c0fae71868f509

**Comment by zou3519** [comment]
> @zou3519 latest update now incorporates functorch's tests plus improvements. True crossref testing I think will late for later.

Awesome, thanks for incorporating those tests!

File: torch/_decomp/decompositions.py [hunk 6]
```diff
@@ -0,0 +1,795 @@
...
+            grad_output / (self * (1.0 - self)),
+            self.new_full((), float("nan")),
+        )
+
+
+@register_decomposition(aten.native_dropout)
+@cast_for_opmath
+def native_dropout_decomposition(input, p, generator=None):
+    bool_mask = torch.rand_like(input) < p
+    res = bool_mask * input * float(1.0 / p)
```
Comment (byronyi) [inline torch/_decomp/decompositions.py:416]: I think these two lines should use (1-p)? cc @ezyang @zou3519 @Chillee @anjali411
